### PR TITLE
[WIP] dockerize readeef

### PR DESCRIPTION
**Pull request by x0rzkov**
Add the following:
- docker-compose file with postgresql, elasticsearch
- dockerfiles for build rd-ng and readeef

### CODE REVIEW

File: .docker/readeef/scripts/entrypoint.sh [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+case "$1" in
+
+  'server')
+  	exec readeef server $@ $ARGS
+	;;
+
+  'client')
+  	exec readeef-client $@ $ARGS
+	;;
```
Comment (urandom) [inline .docker/readeef/scripts/entrypoint.sh]: you can omit that. its a highly experimental webview desktop client. i don't think it has any place in a docker container

**Comment by urandom** [comment]
For some reason, the `ng` in the container does not recognize the `-op` flag. I suggest you change every occurance of that flag in the Makefile to `--output-path`

File: .docker/readeef/scripts/build.sh [hunk 2]
```diff
@@ -0,0 +1,20 @@
+#/bin/sh
+
+set -x
+set -e
+
+# go get github.com/golang/dep/cmd/dep
+# go get github.com/urandom/readeef/cmd/readeef
+go mod tidy
+pwd
```
Comment (urandom) [inline .docker/readeef/scripts/build.sh]: A call to `go get github.com/urandom/embed/cmd/embed` is needed to fetch the embed command

**Comment by x0rzkov** [comment]
> For some reason, the `ng` in the container does not recognize the `-op` flag. I suggest you change every occurance of that flag in the Makefile to `--output-path`
> 
> edit: 2 more ng changes. `--locale` appears to have changed to `--i18n-locale`, and `--i18nFormat` is now `--i18n-format`

i18nFile is fine ? no --i18n-file instead

**Comment by urandom** [comment]
> i18nFile is fine ? no --i18n-file instead

The build passed with the former, though I assume they'll deprecate it eventually, so might as well go with the new --i18n-file flag.

**Comment by urandom** [comment]
I just realized that we don't actually need to build the angular stuff in the container. It should already be generated and available in git. So you don't need npm, or any of that stuff, and you just need to build readeef with go build ./cmd/readeef.

Though the fixes to the makefile would be nice :)

File: .docker/readeef/scripts/entrypoint.sh (hunk 1 above)
Comment (x0rzkov) [reply .docker/readeef/scripts/entrypoint.sh]: removed

**Comment by x0rzkov** [comment]
I created a multi-staged dockerfile, so the last image will be only a tiny one with ./readeef executable.
Also, I fixed the Makefile commands.

# To test with docker
```bash
make docker-build
make docker-run
```


# To test with docker-compose
```bash
docker-compose up
```

**Comment by x0rzkov** [comment]
I have the following error while running readeef with postgresql from docker-compose:

```bash
readeef      | 2019/12/12 13:18:19 Error running server: pq: SSL is not enabled on the server
readeef      | connecting to database
readeef      | github.com/urandom/readeef/content/repo/sql.NewService
readeef      | 	/go/src/github.com/urandom/readeef/content/repo/sql/service.go:28
readeef      | main.runServer
readeef      | 	/go/src/github.com/urandom/readeef/cmd/readeef/server.go:99
readeef      | main.main
readeef      | 	/go/src/github.com/urandom/readeef/cmd/readeef/main.go:74
readeef      | runtime.main
readeef      | 	/usr/local/go/src/runtime/proc.go:203
readeef      | runtime.goexit
readeef      | 	/usr/local/go/src/runtime/asm_amd64.s:1357
readeef      | creating content service
readeef exited with code 1
```

File: docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    links:
+    - db
+    - elasticsearch
+    volumes:
+    - .docker/readeef/config/readeef.toml:/opt/readeef/config/readeef.toml
+    command: ["/opt/readeef/bin/readeef", "-config=/opt/readeef/config/readeef.toml", "server"]
+    #restart: unless-stopped
+
+  rfng:
+    image: rf-ng:alpine3.10-go1.13
```
Comment (urandom) [inline docker-compose.yml]: does this container exist?

File: docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,91 @@
+---
+version: '3.7'
+services:
+
+  readeef:
+    image: readeef:alpine3.10-go1.13
```
Comment (urandom) [inline docker-compose.yml]: looks like this is currently called urandom/readeef

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+# Copy readeef binary to /opt/readeef/bin
+COPY --from=builder /go/src/github.com/urandom/readeef/readeef /opt/readeef/bin/readeef
+COPY .docker/readeef/config/readeef.toml /opt/readeef/config/readeef.toml
+ENV PATH $PATH:/opt/readeef/bin
+
+# Container configuration
+EXPOSE 8080
+VOLUME ["/opt/readeef/data"]
+ENTRYPOINT ["tini", "-g", "--"]
+CMD ["/opt/readeef/bin/readeef", "server"]
```
Comment (urandom) [inline Dockerfile]: the above-copied config isn't currently used. not sure if its an oversight or you're saving it for later.

**Comment by x0rzkov** [comment]
> The docker container itself seems to be building now. Though, if it is possible, its probably better to make it run with postgres when started from compose, and with sqlite (and maybe debug logging) when just running it as a simple container.

That's what I did in fact. I will check other points now

File: Dockerfile (hunk 5 above)
Comment (x0rzkov) [reply Dockerfile]: It is for the sqlite version with only docker mode.

File: docker-compose.yml (hunk 4 above)
Comment (x0rzkov) [reply docker-compose.yml]: I will fix it, did you create an account on hub.docker.com ? or shall I put mine

File: docker-compose.yml (hunk 3 above)
Comment (x0rzkov) [reply docker-compose.yml]: yes, I created a dockerfile to build it. It is in rf-ng dir. As I still struggle to understand how it works I made one anyway. ^^

File: docker-compose.yml (hunk 4 above)
Comment (urandom) [reply docker-compose.yml]: I think I managed to create an org/repo there just now. i think i gave you full permission there

Comment (x0rzkov) [reply docker-compose.yml]: I updated the namespace to unrandom just now.

I have a question about the ui, it seems that we have to copy some files to final container ?
https://github.com/x0rzkov/readeef/blob/dockerize/Dockerfile#L30
https://github.com/x0rzkov/readeef/blob/dockerize/.docker/readeef/config/readeef.toml#L5-L6

Comment (x0rzkov) [reply docker-compose.yml]: Do you have a full config file example ? with all attributes.

Comment (urandom) [reply docker-compose.yml]: > I updated the namespace to unrandom just now.
> 
> I have a question about the ui, it seems that we have to copy some files to final container ?
> https://github.com/x0rzkov/readeef/blob/dockerize/Dockerfile#L30
> https://github.com/x0rzkov/readeef/blob/dockerize/.docker/readeef/config/readeef.toml#L5-L6

we shouldn't need to copy the ui. it's already embedded in the readeef binary.

Comment (x0rzkov) [reply docker-compose.yml]: weird... I have the following login page https://imgur.com/a/IxdeSZP
What are the default credentials ?

**Comment by x0rzkov** [comment]
Let me know what I can do more for this pull request... Normally, it should fit perfectly to your repository :-)

**Comment by urandom** [comment]
> Let me know what I can do more for this pull request...
> Normally, it should fit perfectly to your repository :-)

If its not WIP anymore, I can merge it in

**Comment by x0rzkov** [comment]
please do :-) Have you noticed the dgraph sweet in .docker/dgraph ?

**Comment by x0rzkov** [comment]
wait... I have the following login page https://imgur.com/a/IxdeSZP
What are the default credentials ?

**Comment by urandom** [comment]
The default admin credentials are `admin:admin`

**Comment by x0rzkov** [comment]
when I try to login after docker-compose up, I have the following error from the server.

```bash
readeef          | level=info msg="Getting user admin"
readeef          | level=error msg="gzip handler: http: Handler timeout"
```

**Comment by urandom** [comment]
I can't seem to start the docker-compose setup myself due to errors from iptables:

```
ERROR: for elasticsearch  Cannot start service elasticsearch: driver failed programming external connectivity on endpoint elasticsearch (39110823b1773e8163168a2422803727c71893003cb89e8891be231ddb465dd4):  (iptables failed: iptables --wait -t nat -A DOCKER -p tcp -d 0/0 --dport 9200 -j DNAT --to-destination 172.18.0.2:9200 ! -i br-ac66f82da6fe: iptables v1.8.3 (legacy): unknown option "--to-destination"
```

**Comment by x0rzkov** [comment]
What is your host config ? because it works for me.

Another problem, on my Macbook air, I have the following error while making all without docker.
What are the dependencies (with version) to install to build the ui ?

```bash
readeef % make all
cd rf-ng; ng build --output-path ui/en --base-href /en/ --prod
Schema validation failed with the following errors:
  Data path ".builders['app-shell']" should have required property 'class'.
Error: Schema validation failed with the following errors:
  Data path ".builders['app-shell']" should have required property 'class'.
    at MergeMapSubscriber.project (/usr/local/lib/node_modules/@angular/cli/node_modules/@angular-devkit/core/src/workspace/workspace.js:215:42)
    at MergeMapSubscriber._tryNext (/usr/local/lib/node_modules/@angular/cli/node_modules/rxjs/internal/operators/mergeMap.js:69:27)
    at MergeMapSubscriber._next (/usr/local/lib/node_modules/@angular/cli/node_modules/rxjs/internal/operators/mergeMap.js:59:18)
    at MergeMapSubscriber.Subscriber.next (/usr/local/lib/node_modules/@angular/cli/node_modules/rxjs/internal/Subscriber.js:67:18)
    at MergeMapSubscriber.notifyNext (/usr/local/lib/node_modules/@angular/cli/node_modules/rxjs/internal/operators/mergeMap.js:92:26)
    at InnerSubscriber._next (/usr/local/lib/node_modules/@angular/cli/node_modules/rxjs/internal/InnerSubscriber.js:28:21)
    at InnerSubscriber.Subscriber.next (/usr/local/lib/node_modules/@angular/cli/node_modules/rxjs/internal/Subscriber.js:67:18)
    at MapSubscriber._next (/usr/local/lib/node_modules/@angular/cli/node_modules/rxjs/internal/operators/map.js:55:26)
    at MapSubscriber.Subscriber.next (/usr/local/lib/node_modules/@angular/cli/node_modules/rxjs/internal/Subscriber.js:67:18)
    at SwitchMapSubscriber.notifyNext (/usr/local/lib/node_modules/@angular/cli/node_modules/rxjs/internal/operators/switchMap.js:86:26)
    at InnerSubscriber._next (/usr/local/lib/node_modules/@angular/cli/node_modules/rxjs/internal/InnerSubscriber.js:28:21)
    at InnerSubscriber.Subscriber.next (/usr/local/lib/node_modules/@angular/cli/node_modules/rxjs/internal/Subscriber.js:67:18)
    at /usr/local/lib/node_modules/@angular/cli/node_modules/rxjs/internal/util/subscribeTo.js:17:28
    at Object.subscribeToResult (/usr/local/lib/node_modules/@angular/cli/node_modules/rxjs/internal/util/subscribeToResult.js:10:45)
    at SwitchMapSubscriber._innerSub (/usr/local/lib/node_modules/@angular/cli/node_modules/rxjs/internal/operators/switchMap.js:65:54)
    at SwitchMapSubscriber._next (/usr/local/lib/node_modules/@angular/cli/node_modules/rxjs/internal/operators/switchMap.js:55:14)
make: *** [build-ui-en] Error 1
```

**Comment by x0rzkov** [comment]
https://imgur.com/a/MKlLOsq 

I just added a network layer to docker-compose for elastic
And, I added some feeds successfully.

**Comment by urandom** [comment]
I'm on arch, and the iptables version is: iptables v1.8.3 (legacy)

**Comment by x0rzkov** [comment]
I do not know arch linux but I found the following links about elastic and iptables:
- https://unix.stackexchange.com/questions/256502/elasticsearch-ports-in-iptables
- https://stackoverflow.com/questions/21696667/elasticsearch-restrict-access-using-ip-tables

**Comment by urandom** [comment]
To be honest, I've no idea who calls iptables, though I assume its docker, in order to set up the whatever it is its doing in its network space. That being said, if it works for you, I'll be fine with that, since I said from the start that docker is overkill for my personal usage (though it would've been nice to have a single command startup of readeef + dgraph).

So only thing left is to fix the merge conflict and we can merge it in

**Comment by x0rzkov** [comment]
fs_files.go is generated by embed ? Should we ignore it in .gitignore or what would you advise ?

**Comment by urandom** [comment]
No, the generated file should stay. People shouldn't need to have any of the js stuff in order to be able to use the app

**Comment by urandom** [comment]
Unfortunately, that config file didn't help. Doesn't matter though, if it works for you, it will probably work for other people

**Pull request merged**